### PR TITLE
Resolve cycles using weak mapped identifiers.

### LINKED ISSUE #106: Infinite loop when schema contains a self-reference
```
import * as Enjoi from "https://cdn.skypack.dev/enjoi@8.0.0";

const jsonSchema = {
  type: "object",
  properties: {
    value: { type: "string" },
    next: { $ref: '#' }
  }
}

// add self reference to schema, equivalent to dereferencing `next: { $ref: '#' }`
jsonSchema.properties.next = jsonSchema

const schema = Enjoi.schema(jsonSchema)
const result = schema.validate({
  value: "foo",
  next: {
     value: "bar",
     next: {
       value: 0
     }
  }
})
```

The `result` in the above code sample should return an error about `value: 0` not being a string. Instead Enjoi fails at schema parsing time with a call stack size exceeded error.

This is because Enjoi can't handle schemas with valid self references in them. This issue had been previously raised in #5 and you suggested handling circular references as part of a dereferencing step using another tool. However even when using a fully dereferenced schema, like the above example, Enjoi still breaks.

As is, seems like Enjoi is incapable of parsing valid, recursive JSONSchema no matter how they're passed.

### PR DESCRIPTION
**Pull request by nlundquist**
Resolves #106 

Could I get a review @tlivings ?

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,7 +54,8 @@
   "homepage": "https://github.com/tlivings/enjoi",
   "dependencies": {
     "@hapi/bourne": "^2.0.0",
-    "@hapi/hoek": "^9.0.3"
+    "@hapi/hoek": "^9.0.3",
+    "weak-map": "^1.0.5"
```
Comment (tlivings) [inline package.json:58]: Why is this necessary?

Comment (nlundquist) [reply package.json:58]: can't keep a map with objects as keys without it

Comment (tlivings) [reply package.json:58]: WeakMap is already supported without an additional dependency for a while now.

Comment (nlundquist) [reply package.json:58]: ah I'm running on Node 10

Comment (tlivings) [reply package.json:58]: Should be supported. Supported since 0.12 I think.

Comment (nlundquist) [reply package.json:58]: ah my mistake i read that as Node 12

**Review by tlivings — APPROVED** [review]

**Comment by tlivings** [comment]
Will require resolving conflicts.

**Review by tlivings — APPROVED** [review]
